src/hanabi.sml

```diff
@@ -330,19 +340,21 @@ struct
          hands = tl (#hands s),
          log = getLog (#log s) (length ps),
          turns = #turns s,
+         turnNumber = #turnNumber s,
          inPlay = #inPlay s,
          inDiscard = #inDiscard s}
       val s' = if illegalMove (act,s)
                then raise Fail ("Illegal action: " ^ actionToString act ^ "\n")
                else enact (act,s)
     in
-      if (trace (act,s'); gameOver s')
+      if (trace (act,s); gameOver s')
       then s'
```

Comment:
Did you try my version? It is now:
```
state n
action by the player seeing state n

state n+1
action by the player seeing state n+1

```
This has my preference.

I added printing the end state by adding a `printState s` to `newGame` (which was how you printed the initial state previously). Then the last action isn't printed twice.